File: lab2-functionalprimer/python3/func_as_arg.py
def listFunc():
    return [i for i in range(1, 6)] #Create list of ints from 1 to 5, Haskell equivalent [1..5]

def applicatorFunc(inpFunc, s):
    if s=='s':
        return sum(inpFunc())
    else:
        return sum(inpFunc())/5

print(applicatorFunc(listFunc, 's'))


# Function to generate a list of integers from a to b
def listFunc(a, b):
    return [i for i in range(a, b+1)]  # range is inclusive of b, so we use b+1

# Function to compute either sum or average based on the input 's' or 'a'
def applicatorFunc(inpFunc, a, b, s):
    numbers = inpFunc(a, b)  # Get the list of numbers from inpFunc
    total = sum(numbers)  # Sum of the numbers
    if s == 's':
        return total  # Return sum
    else:
        return total / len(numbers)  # Return average

# Example usage: Compute sum or average for a range of integers
result_sum = applicatorFunc(listFunc, 1, 10, 's')  # Sum of numbers from 1 to 10
print(f"Sum: {result_sum}")

result_avg = applicatorFunc(listFunc, 1, 10, 'a')  # Average of numbers from 1 to 10
print(f"Average: {result_avg}")

File: lab2-functionalprimer/javascript/func_as_arg.js
//Create list of ints from 1 to 5, Haskell equivalent [1..5]
function arrFunc(){
    let arr = [];
    for (let i = 1; i<=5; i++) {
        arr.push(i);
    }
    return arr;    
}

function applicatorFunc(inpFunc, s){
    if(s=='s'){
        const arr = inpFunc();        
        let sum = arr.reduce((accumulator, currentValue) => accumulator + currentValue, 0);
        return sum;
    }
    else{        
        const arr = inpFunc();
        let sum = arr.reduce((accumulator, currentValue) => accumulator + currentValue, 0);
        return sum/5;
    }
}

let x = applicatorFunc(arrFunc, 's');
console.log(x);




// Function to generate a list of integers from a to b
function arrFunc(a, b){
    let arr = [];
    for (let i = a; i <= b; i++) {
        arr.push(i);
    }
    return arr;    
}

// Function to compute either sum or average based on the input 's' or 'a'
function applicatorFunc(inpFunc, a, b, s){
    const arr = inpFunc(a, b);  // Get the array from the function
    let sum = arr.reduce((accumulator, currentValue) => accumulator + currentValue, 0);  // Sum of the array
    if(s == 's'){
        return sum;  // Return sum if 's' is passed
    }
    else {
        return sum / arr.length;  // Return average if 'a' is passed
    }
}

// Example usage: Compute sum or average for a range 1 to 10
let x = applicatorFunc(arrFunc, 1, 10, 's');  // Sum of numbers from 1 to 10
console.log('Sum:', x);

let y = applicatorFunc(arrFunc, 1, 10, 'a');  // Average of numbers from 1 to 10
console.log('Average:', y);
